# What is the *n*th number in the Fibonacci sequence?

The Fibonacci sequence is a sequence of numbers in which each number is the sum of the previous two numers. The sequence begins with 0 and 1. It is defined as 

F<sub>n</sub> = F<sub>n-1</sub> + F<sub>n-2</sub> ,

where 

F<sub>0</sub> = 0 and F<sub>1</sub> = 1.

For example, here is the sequence through the 12th number.

| F<sub>0</sub> | F<sub>1</sub> | F<sub>2</sub> | F<sub>3</sub> | F<sub>4</sub> | F<sub>5</sub> | F<sub>6</sub> | F<sub>7</sub> | F<sub>8</sub> | F<sub>9</sub> | F<sub>10</sub> | F<sub>11</sub> | F<sub>12</sub> |
| :-----------: | :-----------: | :-----------: | :-----------: | :-----------: | :-----------: | :-----------: | :-----------: | :-----------: | :-----------: | :------------: | :------------: | :------------: |
|       0       |       1       |       1       |       2       |       3       |       5       |       8       |      12       |      21       |      34       |       55       |       89       |      144       |

Write a python program to print out the *n*th number in the sequence without using any `for` or `while` loops. 



For example, `fibonacci(10)` should give an output `55`.

For a greater challenge, optimize the above solution to improve runtime complexity. 

## Solution
There are many ways to solve this problem, but if you cannot use loops you must use recursion. Check out this [introductory talk](https://www.youtube.com/watch?v=AfBqVVKg4GE) from North Bay Python con 2018. There are jokes!

In [2]:
def fibonacci(n): 
    if n<0: 
        print("Please stick to positive numbers for now.") 
    # F0 is 0 
    elif n==0: 
        return 0
    # F1 is 1 
    elif n==1: 
        return 1
    else: 
        return fibonacci(n-1)+fibonacci(n-2) 

In [3]:
fibonacci(10)

55

### Bonus!
The complexity of the solution above becomes exponentially intensive. For example, check out the run time below;

In [46]:
%time fibonacci(10)

CPU times: user 37 µs, sys: 286 µs, total: 323 µs
Wall time: 610 µs


55

Can we improve on 610 microseconds? Check out the solution below that uses [memoization](https://en.wikipedia.org/wiki/Memoization), or storage, to become more efficient.

In [43]:
memoize_me_capn = {}
def fibonacci_2(n):
    if n in memoize_me_capn:
        return memoize_me_capn[n]
        
    if n<0:
        print("We are only working on the positive fibonacci numbers for now.")
    elif n == 0:
        value = 0
    elif n == 1:
        value = 1
    else:
        value =  fibonacci_2(n-1)+fibonacci_2(n-2)
    memoize_me_capn[n] = value
    return value

In [44]:
fibonacci_2(10)

55

We can check out the cache.

In [45]:
memoize_me_capn

{1: 1, 0: 0, 2: 1, 3: 2, 4: 3, 5: 5, 6: 8, 7: 13, 8: 21, 9: 34, 10: 55}

Let's see how much computation we saved ourselves.

In [47]:
# clearing the cache!
memoize_me_capn = {}

#now let's compare
%time fibonacci_2(10)

CPU times: user 9 µs, sys: 0 ns, total: 9 µs
Wall time: 11.9 µs


55